src/Features/LanguageServer/Protocol/LanguageServerProtocol.cs

```diff
@@ -1,45 +1,146 @@
+        {
+            _requestHandlers = CreateMethodToHandlerMap(requestHandlers);
+        }
+
+        private static ImmutableDictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>> CreateMethodToHandlerMap(IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        {
+            var requestHandlerDictionary = new Dictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>>();
+            foreach (var lazyHandler in requestHandlers)
+            {
+                requestHandlerDictionary.Add(lazyHandler.Metadata.MethodName, lazyHandler);
+            }
+
+            return requestHandlerDictionary.ToImmutableDictionary();
+        }
 
-        public LanguageServerProtocol(LSP.ClientCapabilities clientCapabilities, bool hierarchicalDocumentSymbolSupport)
+        private async Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(string methodName, Solution solution, RequestType request,
+            LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
         {
-            _hierarchicalDocumentSymbolSupport = hierarchicalDocumentSymbolSupport;
-            _clientCapabilities = clientCapabilities;
+            var handler = (IRequestHandler<RequestType, ResponseType>)_requestHandlers[methodName].Value;
+            return await handler.HandleRequestAsync(solution, request, clientCapabilities, cancellationToken).ConfigureAwait(false);
         }
 
         /// <summary>
-        /// Returns the capabilities of this LSP server.
+        /// Answers an implementation request by returning the implementation location(s) of a given symbol.
+        /// https://microsoft.github.io/language-server-protocol/specification#textDocument_implementation
         /// </summary>
-        public LSP.ServerCapabilities GetServerCapabilities()
+        /// <param name="solution">the solution containing the request document.</param>
+        /// <param name="request">the request document symbol location.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>the location(s) of the implementations of the symbol.</returns>
+        public async Task<object> FindImplementationsAsync(Solution solution, LSP.TextDocumentPositionParams request, LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
         {
-            return new LSP.ServerCapabilities
-            {
-                DefinitionProvider = true,
-                ReferencesProvider = true,
-                DocumentHighlightProvider = true,
-                DocumentSymbolProvider = true,
-                WorkspaceSymbolProvider = true,
-                HoverProvider = true,
-            };
+            return await ExecuteRequestAsync<LSP.TextDocumentPositionParams, object>(LSP.Methods.TextDocumentImplementationName, solution, request, clientCapabilities, cancellationToken)
```

Comment:
all these doesn't need await. basically, if await is used on all exit points (end of method or return), then you can just return task from callee